client/lib/signup/step-actions/index.js

```diff
@@ -283,11 +310,33 @@ export function setThemeOnSite( callback, { siteSlug, themeSlugWithRepo } ) {
 }
 
 export function addPlanToCart( callback, dependencies, stepProvidedItems, reduxStore ) {
-	const { siteSlug } = dependencies;
-	const { cartItem } = stepProvidedItems;
+	const { siteSlug, siteId, domainItem, newSiteParams } = dependencies;
+	const { cartItem, flowName, lastKnownFlow } = stepProvidedItems;
+	const flowToCheck = flowName || lastKnownFlow;
+	const isSiteless = isSitelessCheckoutEnabledForFlow( flowToCheck );
 	if ( isEmpty( cartItem ) ) {
 		// the user selected the free plan
-		defer( callback );
+		if ( isSiteless && siteId === null && isEmpty( domainItem ) && ! isEmpty( newSiteParams ) ) {
```

Comment:
I moved the site creation to a new step called `create-site`, that should make things easier to decouple. 

However in the process I ran into a couple issues:
1) There are several things involved to retrieve the `newSiteParams`, so i moved it to its own function to avoid duplication.
2) The cart updates during the signup are prone to race conditions, if there are several steps being processed at the same time, and both of them try to update the cart, there're high chances one of those updates will be lost, because the current process GETs the cart from the server and then POSTS the new one with the updated products. Right now it's not an issue, but it will probably be in the future, maybe it would make sense to delay the cart creation/updates.